tests/provider/test_in_memory_provider.py

```diff
@@ -0,0 +1,154 @@
+    assert flag.reason == Reason.TARGETING_MATCH
+
+
+def test_should_resolve_boolean_flag_from_in_memory():
+    # Given
+    provider = InMemoryProvider(
+        {"Key": InMemoryFlag("Key", "true", {"true": True, "false": False})}
+    )
+    # When
+    flag = provider.resolve_boolean_details(flag_key="Key", default_value=False)
+    # Then
+    assert flag is not None
+    assert flag.value is True
+    assert isinstance(flag.value, bool)
+    assert flag.variant == "true"
+
+
+def test_should_resolve_integer_flag_from_in_memory():
+    # Given
+    provider = InMemoryProvider(
+        {"Key": InMemoryFlag("Key", "hundred", {"zero": 0, "hundred": 100})}
+    )
+    # When
+    flag = provider.resolve_integer_details(flag_key="Key", default_value=0)
+    # Then
+    assert flag is not None
+    assert flag.value == 100
+    assert isinstance(flag.value, Number)
+    assert flag.variant == "hundred"
+
+
+def test_should_resolve_float_flag_from_in_memory():
+    # Given
+    provider = InMemoryProvider(
+        {"Key": InMemoryFlag("Key", "ten", {"zero": 0.0, "ten": 10.23})}
+    )
+    # When
+    flag = provider.resolve_float_details(flag_key="Key", default_value=0.0)
+    # Then
+    assert flag is not None
+    assert flag.value == 10.23
+    assert isinstance(flag.value, Number)
+    assert flag.variant == "ten"
+
+
+def test_should_resolve_string_flag_from_in_memory():
+    # Given
+    provider = InMemoryProvider(
+        {
+            "Key": InMemoryFlag(
+                "Key",
+                "stringVariant",
+                {"defaultVariant": "Default", "stringVariant": "String"},
+            )
+        }
+    )
+    # When
+    flag = provider.resolve_string_details(flag_key="Key", default_value="Default")
+    # Then
+    assert flag is not None
+    assert flag.value == "String"
+    assert isinstance(flag.value, str)
+    assert flag.variant == "stringVariant"
+
+
+def test_should_resolve_list_flag_from_in_memory():
+    # Given
+    provider = InMemoryProvider(
+        {
+            "Key": InMemoryFlag(
+                "Key", "twoItems", {"empty": [], "twoItems": ["item1", "item2"]}
+            )
+        }
+    )
+    # When
+    flag = provider.resolve_object_details(flag_key="Key", default_value=[])
+    # Then
+    assert flag is not None
+    assert flag.value == ["item1", "item2"]
+    assert isinstance(flag.value, list)
+    assert flag.variant == "twoItems"
+
+
+def test_should_resolve_object_flag_from_in_memory():
+    # Given
+    return_value = {
+        "String": "string",
+        "Number": 2,
+        "Boolean": True,
+    }
+    provider = InMemoryProvider(
+        {"Key": InMemoryFlag("Key", "obj", {"obj": return_value, "empty": {}})}
+    )
+    # When
+    flag = provider.resolve_object_details(flag_key="Key", default_value={})
+    # Then
+    assert flag is not None
+    assert flag.value == return_value
+    assert isinstance(flag.value, dict)
+    assert flag.variant == "obj"
```

Comment:
Good to see tests 🙌 


We need tests for error scenarios. For example, evaluating a flag that's missing (resolved to default with error details) [1], evaluating wrong typed flag (mixing string flag with integer .. etc). Those tests and error handling will make this complete :) 

[1] - https://github.com/open-feature/go-sdk/pull/192/files#diff-60841eaba9b2e1a73e7f57ceebe625fb14c660897d59f7289bd3ce39d05b2342R169-R188

Reply:
1.  Handling default evaluation for missing flags is already tested in `test_should_handle_unknown_flags_correctly`
 2.  The typechecking logic in the Python SDK is implemented in the OpenFeatureClient. Would adding additional typechecking logic in the provider duplicate efforts?

Reply:
1. Thanks for the reference. Connected to comment https://github.com/open-feature/python-sdk/pull/157/files#r1266982206,  we can improve the error handling logic and make this provider behave similar to a real one
2. Interesting, didn't know that this was handled at client level in Python SDK. For example in GO [1] & Java [2] this is not validated at Client. May be it's a good idea to add a simple test and see hoe SDK behaves. Also, this will anyway get covered later by E2E tests 

[1] - https://github.com/open-feature/go-sdk/blob/main/pkg/openfeature/client.go#L709-L732
[2] - https://github.com/open-feature/java-sdk/blob/main/src/main/java/dev/openfeature/sdk/OpenFeatureClient.java#L160C1-L179C1

Reply:
My mistake: there are actually two typechecking errors involved here.

 * On request: when a client attempts to resolve a flag with a method that does not match the flag type
 * On response: when a provider resolves to a flag value that does not match the expected flag type

The response case is checked in OpenFeatureClient which raises a TypeMismatchError.

But the request case should probably be checked by the provider which can either:

 1. Return a FlagResolutionDetails object with the appropriate TYPE_MISMATCH error code.
 2. Raise TypeMismatchError like the OpenFeatureClient

The [specification](https://openfeature.dev/specification/sections/providers#requirement-227) is not clear about which path we should take.